README.rst
=========================
``gs.profile.image.base``
=========================
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Profile images for GroupServer Users
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Author: `Michael JasonSmith`_,
:Contact: Michael JasonSmith <mpj17@onlinegroups.net>
:Date: 2015-05-13
:Organization: `GroupServer.org`_
:Copyright: This document is licensed under a
  `Creative Commons Attribution-Share Alike 4.0 International License`_
  by `OnlineGroups.Net`_.

Introduction
============

This product supplies the code for displaying scaled profile
images. The core scaling is provided by ``gs.image``, while this
product provides

* A `content provider`_ to provide a high-level API, 
* A view_ to display the image at different sizes, 
* A class_ to map a user to an image, and 
* The `missing image`_ resource that is shown if the image cannot
  be found.

Content Provider
================

The ``groupserver.UserImage`` content provider is used by almost all the
code that needs to display a profile image:

.. code-highlight:: xml

  <div tal:define="user view/userInstanceToShow"
       tal:replace="structure provider:groupserver.UserImage" />

It returns a ``<div>`` element (with the ``userimage`` and
``photo`` classes) that contains an ``<img>`` element. The
``src`` attribute of the ``<img>`` element points to the `view`_,
if the user has a profile image, or the `missing image`_
resource. The size of the image defaults to 50×70 pixels
[#units]_ with the ``width`` and ``height`` attributes of the
``<img>`` element set. However, if the longest length of the
image is less that 40 pixels then a `data URI`_ is used to embed
the profile image in the page, rather than using a normal URL.


Two optional arguments can be passed to the content provider to
change the size: ``width`` and ``height``. While the size of the
image will be changed, the aspect ratio of the image is preserved
(see View_ below):

.. code-highlight:: xml

  <div tal:define="user view/userInstanceToShow;
                   width string:14; height string:18;"
       tal:replace="structure provider:groupserver.UserImage" />

View
====

The view ``/p/{userId}/gs-profile-image`` is a view that returns
the image for the user. The ``gs-profile-image`` view provides an
**API** for viewing the profile image at different sizes
[#api]_. Two optional arguments can be passed as part of a path,
to set the width and the height of the image.

* Not providing any arguments displays the **original** image
  that the participant uploaded:
  <http://groupserver.org/p/mpj17/gs-profile-image>.

* If only one argument is provided, the **width** is set and the
  height will be calculated, keeping the same aspect ratio as the
  original image. A new image will be returned that is scaled as
  requested. For example, setting the width to 100px:
  <http://groupserver.org/p/mpj17/gs-profile-image/100>.

* The **height** is specified after the width. The image will be
  scaled so neither the width nor the height will be
  exceeded. However, the *aspect ratio* will be preserved, so one
  dimension may be smaller than requested [#resize]_. For
  example, ensuring that the width does not exceed 100px, and the
  height is smaller than to 125px:
  <http://groupserver.org/p/mpj17/gs-profile-image/100/125>.

The scaling of the image is carried out by the ``UserImage``
class_.  If the participant lacks an image then the viewer is
redirected to the `missing image`_ resource instead.

Class
=====

The class ``gs.profile.image.base.UserImage`` is a subclass of
``gs.image.GSImage``. It differs in the constructor, which takes
a ``context`` and a user-info, rather than a file.

The ``context`` and user-info are used to construct a *glob* for
the file [#glob]_, looking in the ``groupserver.user.image``
data-directory (usually found beneath
``var/instance/groupserver.data`` in the GroupServer
directory). A glob is used so different file types, with
different extensions, can be used.

If the file cannot be found then an ``IOError`` is thrown. The
error has the following properties set.

``errno``:
  The error number is set to ``errno.ENOENT``.

``filename``:
  The *glob* that was used to try and find the file.

If the file is found then it is opened and passed to the
constructor for ``GSImage``.

Missing Image
=============

The resource ``/++resource++gs-profile-image-base-missing.jpg``
is the *missing profile image* image.

Resources
=========

- Code repository:
  https://github.com/groupserver/gs.profile.image.base
- Questions and comments to
  http://groupserver.org/groups/development
- Report bugs at https://redmine.iopen.net/projects/groupserver

.. _GroupServer: http://groupserver.org/
.. _GroupServer.org: http://groupserver.org/
.. _OnlineGroups.Net: https://onlinegroups.net/
.. _Michael JasonSmith: http://groupserver.org/p/mpj17/
.. _Creative Commons Attribution-Share Alike 3.0 New Zealand License:
   http://creativecommons.org/licenses/by-sa/3.0/nz/

.. [#units] 50×70 pixels is 2.5×3.5 units in the standard
            `GroupServer CSS`_.
.. _GroupServer CSS: https://github.com/groupserver/gs.content.css

.. _data URI: http://tools.ietf.org/html/rfc2397
.. [#api] The API is the same as the images supported by
          ``Products.XWFFileLibrary2``.

.. [#resize] See ``gs.image`` for details on resizing
             <https://github.com/groupserver/gs.image>

.. [#glob] See `the glob module`_.
.. _the glob module: http://docs.python.org/2.7/library/glob.html


docs/HISTORY.rst
Changelog
=========

2.2.1 (2015-05-13)
------------------

* Switching to Unicode literals
* Naming the reStructuredText files as such
* Pointing at GitHub_ as the primary repository

.. _GitHub: https://github.com/groupserver/gs.profile.image.base

2.2.0 (2014-01-23)
------------------

* Adding more to the ``__init__.py`` for the product
* Encoding a URL as ASCII

2.1.0 (2013-07-15)
------------------

* Fixing a Unicode error
* Changing the definition of the ``showImage`` attribute of the
  profile

2.0.0 (2013-04-03)
------------------

* Using a ``data:`` URI for small images

1.1.0 (2013-02-26)
------------------

* Changing the default size of the image
* Checking if the image exists

1.0.0 (2012-12-07)
------------------

Initial version. Prior to the creation of this product the
profile image was handled by `Products.GSProfile`_.

.. _Products.GSProfile: https://github.com/groupserver/Products.GSProfile


version.py
# -*- coding: utf-8 -*-
version = '2.2.1'
release = False

#--------------------------------------------------------------------------#
import sys
if (sys.version_info < (3, )):
    from commands import getstatusoutput
else:
    from subprocess import getstatusoutput  # lint:ok
import datetime
import os
import glob


class CommandError(Exception):
    pass


def execute_command(commandstring):
    status, output = getstatusoutput(commandstring)
    if status != 0:
        m = 'Command "{0}" exited with status {1}'
        msg = m.format(commandstring, status)
        raise CommandError(msg)
    return output


def parse_version_from_package():
    try:
        pkginfo = os.path.join(glob.glob('*.egg-info')[0],
                               'PKG-INFO')
    except:
        pkginfo = ''

    version_string = ''
    if os.path.exists(pkginfo):
        for line in open(pkginfo):
            if line.find('Version: ') == 0:
                version_string = line.strip().split('Version: ')[1].strip()
        if not version_string:
            version_string = '%s-dev' % version
    else:
        version_string = version

    return version_string


def get_version():
    try:
        globalid = execute_command("hg identify -i").strip('+')
        c = "hg log -r %s --template '{date|isodatesec}'" % globalid
        commitdate = execute_command(c)
        # convert date to UTC unix timestamp, using the date command because
        # python date libraries do not stabilise till about 2.6
        dateCommand = 'date -d"%s" --utc +%%s' % commitdate
        timestamp = int(execute_command(dateCommand))

        # finally we have something we can use!
        dt = datetime.datetime.utcfromtimestamp(timestamp)
        datestring = dt.strftime('%Y%m%d%H%M%S')
        if release:
            version_string = version
        else:
            version_string = "%s.dev%s-%s" % (version, datestring, globalid)

    except (CommandError, ValueError, TypeError):
        # --=mpj17=-- Usually because we are building out a source-egg,
        # rather than from a Hg source-directory.
        version_string = parse_version_from_package()

    return version_string

if __name__ == '__main__':
    import sys
    sys.stdout.write('{0}\n'.format(get_version()))


gs/profile/image/base/image.py
# -*- coding: utf-8 -*-
############################################################################
#
# Copyright © 2014, 2015 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
############################################################################
from __future__ import absolute_import, unicode_literals
from zope.cachedescriptors.property import Lazy
from gs.core import to_ascii
from gs.profile.base.page import ProfilePage
from .userimage import UserImage


class Image(ProfilePage):

    def __init__(self, context, request):
        super(Image, self).__init__(context, request)
        self.traverse_subpath = []

    def publishTraverse(self, request, name):
        self.traverse_subpath.append(name)
        return self

    @Lazy
    def userImage(self):
        retval = UserImage(self.context, self.userInfo)
        return retval

    @Lazy
    def width(self):
        tsp = self.traverse_subpath
        if len(tsp) > 0:
            retval = int(tsp[0])
        else:
            retval = self.userImage.width
        return retval

    @Lazy
    def height(self):
        tsp = self.traverse_subpath
        if len(tsp) >= 2:
            retval = int(tsp[1])
        elif len(tsp) == 1:
            r = float(self.userImage.height) / float(self.userImage.width)
            retval = int((self.width * r) + 0.5)
        else:
            retval = self.userImage.height
        return retval

    @Lazy
    def image(self):
        if self.traverse_subpath:
            retval = self.userImage.get_resized(self.width, self.height,
                                                maintain_aspect=True,
                                                only_smaller=False)
        else:
            retval = self.userImage
        return retval

    def __call__(self):
        # TODO: Add uWSGI offloading with X-Sendfile
        # http://uwsgi-docs.readthedocs.org/en/latest/articles/OffloadingWebsocketsAndSSE.html#uwsgi-offloading
        try:
            h = 'inline; filename={0}-{1}x{2}.jpg'
            hdr = h.format(self.userInfo.nickname, self.width, self.height)
            self.request.RESPONSE.setHeader(b'Content-Disposition',
                                            to_ascii(hdr))

            self.request.RESPONSE.setHeader(b'Cache-Control',
                                            b'private; max-age=1200')

            self.request.RESPONSE.setHeader(
                b'Content-Type', to_ascii(self.image.contentType))

            self.request.RESPONSE.setHeader(
                b'Content-Length', to_ascii(str(self.image.getSize())))
            retval = self.image.data
        except IOError:
            missingImage = b'/++resource++gs-profile-image-base-missing.jpg'
            retval = self.request.RESPONSE.redirect(missingImage)
        return retval


gs/profile/image/base/userimage.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
from errno import ENOENT
from glob import glob
import os.path
from gs.image import GSImage
from Products.XWFCore.XWFUtils import locateDataDirectory


class UserImage(GSImage):

    def __init__(self, context, userInfo):
        assert context
        self.context = context

        assert userInfo
        self.userInfo = userInfo

        self.file = get_file(context, userInfo)
        super(UserImage, self).__init__(self.file)


def get_file(context, userInfo):
    retval = None
    imagePath = get_image_path(context, userInfo)
    if imagePath:
        retval = file(imagePath, 'rb')
    return retval


def get_image_path(context, userInfo):
    # TODO: Cache
    # --=mpj17=-- Note to Future Coder: version numbers could be added to
    # the files: something like userId-YYYYMMDDHHMMSS.ext ?
    # '{0}-[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]'
    # '[0-9][0-9].*'.format(self.userInfo.id)
    filename = '{0}.*'.format(userInfo.id)
    imagePath = os.path.join(get_imageDir(context), filename)

    retval = None
    files = glob(imagePath)
    if files and os.path.isfile(files[0]):
        retval = files[0]
    else:
        m = u'Cannot open the profile image for {name} ({id})'
        msg = m.format(name=userInfo.name, id=userInfo.id)
        raise IOError(ENOENT, msg, imagePath)
    return retval


def get_imageDir(context):
    # TODO: Cache
    site_root = context.site_root()
    siteId = site_root.getId()
    retval = locateDataDirectory("groupserver.user.image", (siteId,))
    return retval


gs/profile/image/base/contentprovider.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
from __future__ import absolute_import
from base64 import b64encode
from zope.cachedescriptors.property import Lazy
from zope.contentprovider.interfaces import UpdateNotCalled
from zope.pagetemplate.pagetemplatefile import PageTemplateFile
from Products.CustomUserFolder.interfaces import IGSUserInfo
from Products.CustomUserFolder.userinfo import GSAnonymousUserInfo
from gs.viewlet.contentprovider import SiteContentProvider
from .userimage import UserImage


class UserImageContentProvider(SiteContentProvider):

    def __init__(self, context, request, view):
        super(UserImageContentProvider, self).__init__(context, request, view)
        self.updated = False

    def update(self):
        self.updated = True
        self.pageTemplate = PageTemplateFile(self.pageTemplateFileName)
        self.finalSize = self.get_final_size()
        self.finalWidth = self.finalSize[0]
        self.finalHeight = self.finalSize[1]
        self.imageUrl = self.get_image_url()

    def render(self):
        if not self.updated:
            raise UpdateNotCalled
        return self.pageTemplate(view=self)

    def get_final_size(self):
        if self.showMissingImage:
            retval = [int(d) for d in (self.width, self.height)]
        elif self.resizeNeeded:
            retval = self.smallImage.getImageSize()
        else:
            retval = self.userImage.getImageSize()
        return retval

    @Lazy
    def showMissingImage(self):
        try:
            retval = (self.userInfo.anonymous or (self.userImage.file is None)
                        or not(getattr(self.user, 'showImage', False)))
        except IOError:
            retval = True
        return retval

    @Lazy
    def resizeNeeded(self):
        return ((int(self.width) < self.userImage.width) or
                    (int(self.height) < self.userImage.height))

    @Lazy
    def smallImage(self):
        if self.resizeNeeded:
            retval = self.userImage.get_resized(int(self.width),
                                                    int(self.height))
        else:
            retval = self.userImage
        return retval

    @Lazy
    def userImage(self):
        return UserImage(self.context, self.userInfo)

    @Lazy
    def userInfo(self):
        try:
            retval = IGSUserInfo(self.user)
        except TypeError:
            retval = GSAnonymousUserInfo()
        return retval

    def get_image_url(self):
        retval = self.missingImage  # From the interface
        try:
            if not self.showMissingImage:
                if self.smallImage.getSize() < 1023:
                    retval = self.embedded_profile_image()
                elif self.resizeNeeded:
                    retval = self.resize_link()
                else:
                    retval = self.profile_image_link()
        except IOError:
            pass  # Use the missingImage
        return retval

    def profile_image_link(self):
        r = '{profile}/gs-profile-image'
        return r.format(profile=self.userInfo.url)

    def resize_link(self):
        r = '{profileLink}/{width}/{height}'
        return r.format(profileLink=self.profile_image_link(),
                            width=self.width, height=self.height)

    def embedded_profile_image(self):
        d = b64encode(self.smallImage.data)
        r = 'data:{mediatype};base64,{data}'
        return r.format(mediatype=self.smallImage.contentType, data=d)


gs/profile/image/base/interfaces.py
# -*- coding: utf-8 -*-
##############################################################################
#
# Copyright © 2014 OnlineGroups.net and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
from __future__ import unicode_literals
from zope.contentprovider.interfaces import IContentProvider
from zope.schema import ASCIILine, Bool, Field, Int

# FIXME: use gs.config
MISSING_IMAGE_URL = '/++resource++gs-profile-image-base-missing.jpg'


class IGSUserImage(IContentProvider):
    """User Image"""
    pageTemplateFileName = ASCIILine(
        title="Page Template File Name",
        description='The name of the ZPT file that is used to render the '
                    'profile image.',
        required=False,
        default="browser/templates/userimage.pt".encode('ascii', 'ignore'))

    user = Field(
        title='User Instance',
        description='An instance of the CustomUser Class',
        required=True)

    showImageRegardlessOfUserSetting = Bool(
        title='Show Image Regardles of User Setting',
        description="Show the user's image, regardless of the value of "
                    "the showImage property. This should be used with "
                    "extreme caution, as it can violate the user's privacy.",
        required=False,
        default=False)

    width = Int(
        title='Width',
        description='The width of the image, in pixels.',
        required=False,
        default=50)  # FIXME: use gs.config

    height = Int(
        title='Height',
        description='The height of the image, in pixels.',
        required=False,
        default=70)  # FIXME: use gs.config

    missingImage = ASCIILine(
        title='Missing Image',
        description='The URL of the image to use for the missing-image '
                    'image.',
        required=False,
        default=MISSING_IMAGE_URL.encode('ascii', 'ignore'))
